### PR TITLE
Assignment 1. Implement additional logic and random card theme

### PR DESCRIPTION
**Pull request by Shiaulis**
- added more cards
- added "new game" button
- added shuffle logic
- moved flip counter logic to the game
- added score logic and label that reflects actual score
- added card themes
- added card back color to theme

### CODE REVIEW

File: Concentration/Concentration.swift [hunk 1]
```diff
@@ -6,24 +6,39 @@
...
 
-    var cards: [Card] = []
+    private (set) var cards: [Card] = []
+    private (set) var flipCount: Int = 0
+    private (set) var score: Int = 0
+
+    let cardBackgroundColor: UIColor
+
+    private var emojiChoices: [String]
+    private var emoji = [Int: String]()
```
Comment (Shiaulis) [inline Concentration/Concentration.swift:24]: for me this way of setting/getting data is questionable. However I didn't want to refactor initial lecturer logic so decided to use the same way of setting/getting emojis

File: Concentration/Concentration.swift [hunk 2]
```diff
@@ -36,13 +51,27 @@ final class Concentration {
         if let matchIndex = self.indexOfOneAndOnlyFaceUpCard, matchIndex != chosenIndex {
             if isMatching(at: [chosenIndex, matchIndex]) {
                 markAsMatched(at: [chosenIndex, matchIndex])
+                self.score += 2
```
Comment (Shiaulis) [inline Concentration/Concentration.swift:54]: not sure if it worth to make a separate method for scoring…

File: Concentration/Concentration.swift [hunk 3]
```diff
@@ -36,13 +51,27 @@ final class Concentration {
...
         }
         else {
             turnFaceDownAll()
         }
 
         turnFaceUp(at: chosenIndex)
+        self.flipCount += 1
+    }
+
+    func generateEmoji(for card: Card) -> String {
```
Comment (Shiaulis) [inline Concentration/Concentration.swift:68]: kinda the same method that was initially written in view controller

File: Concentration/ConcentrationTheme.swift [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+            return .systemGray
+        case .fruits:
+            return .systemOrange
+        case .vehicles:
+            return .systemRed
+        }
+    }
+
+    static func getRandomTheme() -> ConcentrationTheme {
+        ConcentrationTheme.allCases.randomElement() ?? .animals
```
Comment (Shiaulis) [inline Concentration/ConcentrationTheme.swift:46]: have no idea why `randomElement` return value is optional 🤷🏾

File: Concentration/ViewController.swift [hunk 5]
```diff
@@ -11,58 +11,54 @@ final class ViewController: UIViewController {
 
     // MARK: - Properties -
 
-    private static var emojiChoices = ["🦇", "😱", "🙀", "😈", "🎃", "👻", "🍭", "🍬", "🍎"]
+    private lazy var game = Concentration(numberOfCards: self.cardButtons.count, theme: .allCases.randomElement() ?? .animals)
```
Comment (Shiaulis) [inline Concentration/ViewController.swift]: I assume we could use some better way of setting new game as we did in Matchismo however I was too lazy (pun intended) to do that

File: Concentration/ViewController.swift [hunk 6]
```diff
@@ -11,58 +11,54 @@ final class ViewController: UIViewController {
...
-            print("chosen card was not in cardButtons")
+            assertionFailure("chosen card was not in cardButtons")
         }
     }
 
+    // MARK: - Lifecycle -
+
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        updateViewFromModel()
```
Comment (Shiaulis) [inline Concentration/ViewController.swift:43]: otherwise initial card back theme won't be set

File: Concentration/Concentration.swift [hunk 7]
```diff
@@ -6,24 +6,39 @@
...
+import UIKit
 
 final class Concentration {
 
     typealias CardIndex = Array<Card>.Index
 
     // MARK: - Properties -
 
-    var cards: [Card] = []
+    private (set) var cards: [Card] = []
```
Comment (khaletska) [inline Concentration/Concentration.swift:17]: what's this?

File: Concentration/Concentration.swift (hunk 2 above)
Comment (khaletska) [reply Concentration/Concentration.swift:54]: I don't think so 😊

File: Concentration/ViewController.swift (hunk 5 above)
Comment (khaletska) [reply Concentration/ViewController.swift]: move it to viewDidLoad?

Comment (khaletska) [reply Concentration/ViewController.swift]: and also you have getRandomTheme

**Review by khaletska — CHANGES_REQUESTED** [review]
well I feel like themes are overcomplicated but let's just move the initialiser of the game so it is not repeated and that's it

File: Concentration/Concentration.swift (hunk 7 above)
Comment (Shiaulis) [reply Concentration/Concentration.swift:17]: this is a property that public for getting but private for setting. Basically, readonly

**Review by khaletska — APPROVED** [review]
Niiiice 🍀

**Pull request merged**